crates/bevy_ui/src/render/mod.rs

```diff
@@ -320,48 +321,42 @@ pub fn extract_text_uinodes(
         if let Ok((uinode, global_transform, text, text_layout_info, visibility, clip)) =
             uinode_query.get(*entity)
         {
-            if !visibility.is_visible() {
-                continue;
-            }
-            // Skip if size is set to zero (e.g. when a parent is set to `Display::None`)
-            if uinode.size() == Vec2::ZERO {
+            // Skip if not visible or if size is set to zero (e.g. when a parent is set to `Display::None`)
+            if !visibility.is_visible() || uinode.size().x == 0. || uinode.size().y == 0. {
                 continue;
             }
-            let text_glyphs = &text_layout_info.glyphs;
-            let alignment_offset = (uinode.size() / -2.0).extend(0.0);
 
             let mut color = Color::WHITE;
             let mut current_section = usize::MAX;
-            for text_glyph in text_glyphs {
-                if text_glyph.section_index != current_section {
-                    color = text.sections[text_glyph.section_index]
-                        .style
-                        .color
-                        .as_rgba_linear();
-                    current_section = text_glyph.section_index;
+            let scale = scale_factor.recip();
+            let transform = global_transform.compute_matrix()
+                * mat4(
+                    Vec4::X * scale,
+                    Vec4::Y * scale,
+                    Vec4::Z * scale,
```

Comment:
Is there no simpler way to make this work? Not that this is much harder to read than the dense matrix math we had before, but it definitely is still quite opaque in what it's doing.

Separately, not sure if this actually does vector multiplication here or it just initializes a single component to the value. Probably doesn't make that big of a difference, but I wonder if we're leaving some extra perf on the table.

Reply:
> Is there no simpler way to make this work? Not that this is much harder to read than the dense matrix math we had before, but it definitely is still quite opaque in what it's doing.

First draft of this PR I used:
```
transform.compute_matrix()
        * Mat4::from_scale_rotation_translation(
            Vec3::splat(scale),
            Quat::IDENTITY,
            -0.5 * uinode.size().extend(0.),
        )
```
But I didn't like including the rotation and replaced it with defining the columns explicitly. Looking at it again though it seems that the rotation is removed with full optimizations, which must be why glam has no `from_scale_translation` function I guess. There's no advantage here, so I'll change it back.

> Separately, not sure if this actually does vector multiplication here or it just initializes a single component to the value. Probably doesn't make that big of a difference, but I wonder if we're leaving some extra perf on the table.

The output generated by ` Vec4::X * scale` and `Vec4 { x: scale, .. Vec4::ZERO }` is identical, as far as I can tell.

Reply:
I understand better now, just leave everything to the compiler. This is much nicer:
```rust
let transform = global_transform.compute_matrix()
                * Mat4::from_translation(-0.5 * uinode.size().extend(0.))
                * Mat4::from_scale(Vec3::splat(scale_factor.recip()));
```
It even seems to give a slight performance improvement, though I'm not sure why.